subprojects/snapshots/src/main/java/org/gradle/internal/vfs/impl/DefaultSnapshotHierarchy.java

```diff
@@ -84,6 +86,15 @@ public SnapshotHierarchy empty() {
         return empty(caseSensitivity);
     }
 
+    @Override
+    public void visitCompleteSnapshots(Consumer<CompleteFileSystemLocationSnapshot> snapshotVisitor) {
+        rootNode.accept(node -> {
+            if (node instanceof CompleteFileSystemLocationSnapshot) {
```

Comment:
How about using `node.getSnapshot().ifPresent()`? That would be more future proof I think.

Reply:
`MetadataSnapshot` doesn't include the absolute path unfortunately.

Reply:
I know, you still need to do the `instanceof` check. Though I think we want to move away from node being snapshots.

Reply:
Good point, added.

Reply:
Would renaming `MetadataSnapshot` to `FileMetadata` and `CompleteFileLocationSnapshot` to `FileLocationSnapshot` clarify things? (I'd love to rename the latter to `FileSystemSnapshot`, but that's still used elsewhere.)

Reply:
I am not sure that will help. How about we take all the names we have next week and try to come up with some consistent naming now that we have a better understanding?

Reply:
👍